securesystemslib/signer.py

```diff
@@ -137,8 +150,264 @@ def to_dict(self) -> Dict:
+
+        key_impl = KEY_FOR_TYPE_AND_SCHEME[(keytype, scheme)]
+        return key_impl.from_dict(keyid, key_dict)
+
+    def to_dict(self) -> Dict[str, Any]:
+        """Returns a dict for TUF serialization.
+
+        Key implementations may override this method.
+        """
+        return {
+            "keytype": self.keytype,
+            "scheme": self.scheme,
+            "keyval": self.keyval,
+            **self.unrecognized_fields,
+        }
+
+    @abc.abstractmethod
+    def verify_signature(self, signature: Signature, data: bytes) -> None:
+        """Verifies the signature over data.
+
+        Args:
+            signature: Signature object.
+            data: Payload bytes.
+
+        Raises:
+            UnverifiedSignatureError: Failed to verify signature.
+            VerificationError: Signature verification process failed. If you
+                are only interested in the verify result, just handle
+                UnverifiedSignatureError: it contains VerificationError as well
+        """
+        raise NotImplementedError
+
+
+class SSlibKey(Key):
+    """Key implementation for RSA, Ed25519, ECDSA and Sphincs keys"""
+
+    def to_securesystemslib_key(self) -> Dict[str, Any]:
+        """Internal helper function"""
+        return {
+            "keyid": self.keyid,
+            "keytype": self.keytype,
+            "scheme": self.scheme,
+            "keyval": self.keyval,
+        }
+
+    @classmethod
+    def from_securesystemslib_key(cls, key_dict: Dict[str, Any]) -> "SSlibKey":
+        """Constructor from classic securesystemslib keydict"""
+        return SSlibKey(
+            key_dict["keyid"],
+            key_dict["keytype"],
+            key_dict["scheme"],
+            key_dict["keyval"],
+        )
+
+    @classmethod
+    def from_dict(cls, keyid: str, key_dict: Dict[str, Any]) -> "SSlibKey":
+        keytype = key_dict.pop("keytype")
+        scheme = key_dict.pop("scheme")
+        keyval = key_dict.pop("keyval")
+        # All fields left in the key_dict are unrecognized.
+        return cls(keyid, keytype, scheme, keyval, key_dict)
+
+    def verify_signature(self, signature: Signature, data: bytes) -> None:
+        try:
+            if not sslib_keys.verify_signature(
+                self.to_securesystemslib_key(),
+                signature.to_dict(),
+                data,
+            ):
+                raise exceptions.UnverifiedSignatureError(
+                    f"Failed to verify signature by {self.keyid}"
+                )
+        except (
+            exceptions.CryptoError,
+            exceptions.FormatError,
+            exceptions.UnsupportedAlgorithmError,
+        ) as e:
+            logger.info("Key %s failed to verify sig: %s", self.keyid, str(e))
+            raise exceptions.VerificationError(
+                f"Unknown failure to verify signature by {self.keyid}"
+            ) from e
+
+
+@dataclass
+class GPGKey(Key):  # pylint: disable=too-many-instance-attributes
```

Comment:
I feels odd that `GPGKey` is a `Key` although it has different fields and cannot be instantiated with `Key.from_dict`. IMO it makes only sense to implement this in `securesystemslib` if it is relevant for `python-tuf` and DSSE as well, otherwise we should just implement it in `in-toto`.

Reply:
Key.from_dict does work for GPGKey -- as long as GPGKey is in the supported keys list.

I don't think the current GPGKey will ever be relevant to python-tuf: both keys and signatures break the spec and we can't change that because in-toto depends on this behaviour. I can't imagine the spec changing to allow current GPGPKey behaviour but I suppose that could happen.

But I don't think there is a problem with
* keeping GPGKey in securesystemslib as long as intoto uses it
* if there is demand, adding a compliant gpg key version (using a different Key and Signer subclass)

Reply:
> Key.from_dict does work for GPGKey -- as long as GPGKey is in the supported keys list, see test_key_from_to_dict().

Oh right. But do we really want to tell users to add `(None, None)` to the dispatch table?  IMO it's okay for a user of the `Key` interface to do that as a workaround for a custom `Key` subclass. But we should not require it for a `Key` subclass that we provide. It feels like an abuse of our own API.

Reply:
Yes, GPG support as it is, is certainly a hack. 

Moving the (Keys+Signer) GPG support to in-toto sounds good to me. We just need to make sure securesystemslib allows whatever is needed to do it:
* out-of-tree Key and Signer
* Key dispatch that somehow allows dispatching to GPGKey even though it's a bit different
* keyid matching has to be made to work with GPG (this is not an issue wright now but may be with DSSE) 

I think this all looks solved right now, at least well enough

Reply:
wait, I maybe misunderstood your comment, you actually mention a workaround within Key... I'm not enthusiastic on paper but we can discuss this when we next talk

Reply:
I meant that it's okay for in-toto to implement `GPGKey` without `keytype` and `scheme` and register it under `securesystemslib.signer.KEY_FOR_TYPE_AND_SCHEME[(None, None)]` as a workaround.  But it seems odd if securesystemslib implements `GPGKey` and requires in-toto to perform that workaround.

I think we are on the same page.

Reply:
Just to document this for others: I am removing GPGKey from the PR. The idea is that the current GPG implementation likely moves to intoto.